### PR TITLE
Add support for supervisor fun to Task.Supervisor.async_stream

### PR DESCRIPTION
**Pull request by antipax**
This PR allows you to pass a 1-arity fun as the `supervisor` argument for the `async_stream` functions, which receives the stream entry and should return a supervisor pid/name, which will then be used as the target supervisor for the task which will process the stream entry.

cc: @josevalim @ericmj

### CODE REVIEW

File: lib/elixir/test/elixir/task/supervisor_test.exs [hunk 1]
```diff
@@ -22,6 +22,16 @@ defmodule Task.SupervisorTest do
     number
   end
 
+  def sleep_and_return_ancestor(number, _another_arg \\ nil) do
+      Process.sleep(number)
```
Comment (josevalim) [inline lib/elixir/test/elixir/task/supervisor_test.exs]: Wrong indentation? And maybe this function could be private?

File: lib/elixir/test/elixir/task/supervisor_test.exs [hunk 2]
```diff
@@ -22,6 +22,16 @@ defmodule Task.SupervisorTest do
     number
   end
 
+  def sleep_and_return_ancestor(number, _another_arg \\ nil) do
```
Comment (josevalim) [inline lib/elixir/test/elixir/task/supervisor_test.exs]: Should we rather implement this as:

```elixir
defp sleep_and_return_ancestors(number, :another_arg) do
  sleep_and_return_ancestors(number)
end

defp sleep_and_return_ancestors(number) do
  # code today
end
```

So we assert the argument is correct?

**Comment by josevalim** [comment]
I have added some comments but it looks good to go!

**Comment by antipax** [comment]
Sorry about the indentation, dunno how I missed that. :)

I don't think it can be private, because we use it to test the MFA form of async_stream (vs. the fun version), but it's definitely better to assert on the arg, so I've updated that. Thanks!

**Comment by whatyouhide** [comment]
The implementation looks good to me, but I have questions about the use case behind this: do we have some good use cases that justify this? If so, can we document at least one of them through an example when talking about using a function as the supervisor?

**Comment by antipax** [comment]
Sure, it's intended use case is distributing the tasks amongst multiple nodes in a distributed environment. I'll add that example to the docs, if that works?

**Comment by josevalim** [comment]
@antipax please do! Documentation on use cases is always good.

**Comment by josevalim** [comment]
@fishcakez I think fun is executed in the caller process, in any case it is a great idea to test it. And if it is in the same process as the caller, we should also test and document it.

**Comment by antipax** [comment]
@fishcakez sure, it could definitely be more general than the single use case I had in mind, thanks for taking a look! Do you think I should further update the documentation around the possible use cases beyond what I just pushed?

@josevalim it is indeed executed in the caller process, and I agree it is a useful fact to add the documentation

I will take the extra testing feedback from both of you and implement it very soon, thanks for the feedback everyone

**Comment by whatyouhide** [comment]
@antipax so to understand, we should have more docs on possible use cases, and mention that this function is executed in the caller process, right?

**Comment by josevalim** [comment]
Ping!

**Comment by antipax** [comment]
Sorry, yep, know what needs to be done here and have just been real busy! :) Planning to tackle this tomorrow or later today.

**Comment by antipax** [comment]
When I wrote these tests I found out that I was actually wrong before. It doesn't execute in the caller process, it executes in the monitor process that is responsible for spawning the processes and translating down messages for the caller process. I have written the docs to explain that it is executed in a separate process (while trying not to go into too much detail about internal implementation details) but I wonder if anyone thinks more should be added?

Ultimately, the failures that occur are the same sort as if you provide an invalid or dead supervisor normally to both the linked and unlinked async streams, so it seems to me like perhaps we could add additional documentation for that fact, or leave it as is. What does everyone think?

**Comment by josevalim** [comment]
This looks good to me, thank you!

**Comment by whatyouhide** [comment]
Thanks @antipax! 💟

**Comment by fishcakez** [comment]
We may need to look into the error handling of the `supervisor_fun`. Particularly around how we are handling the errors. Previously if the supervisor crashed the middleman would also crash when it tried to start a child. However in that scenario the supervisor will shutdown the tasks when it terminates.

Unfortunately by allowing user land code in this middleman process we lose this guarantee because the `supervisor_fun` can crash and the take the middle man with it but the tasks won't be shutdown by the supervisor. This loses the guarantee when trapping exits in the caller (and so middleman) that tasks will be stopped. In a evenly less likely scenario if the `supervisor_fun` calls `exit(:normal)` we leak the tasks no matter what :(.

The high level issue we have is that `supervisor_fun` is allowing user code to run in a process that is doing supervisor and/or scheduler role. This is an anti-pattern because a bug in the user logic can prevent guarantees of the process.

I think we should consider moving this to a stream map on the calling process and/or adding more tests to verify behavior when a crash occurs.

**Comment by josevalim** [comment]
@fishcakez should we catch any error that happen in the supervised fun and send it downstream and re-raise from there?

**Comment by josevalim** [comment]
> The easiest test to make sure we get it right is to call exit(: normal) in second event supervisor fun and check we cleanup first event task

**Comment by antipax** [comment]
Happy to add tests, as well as update the code once we decide on a path.
Seems like the two options are either running the supervisor fun on the
caller process, or propagating the error to the caller, if testing reveals
tasks aren't cleaned up? By the way, when we say cleaned up, what does that
mean? Is it that they're sent an exit to cancel any in-progress work?

On Oct 7, 2017 3:56 PM, "José Valim" <notifications@github.com> wrote:

The easiest test to make sure we get it right is to call exit(: normal) in
second event supervisor fun and check we cleanup first event task

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/elixir-lang/elixir/pull/6552#issuecomment-334961870>,
or mute the thread

**Comment by josevalim** [comment]
Moving it to the caller process is a major change. It is preferrable to guarantee proper clean-up in the middleman.

**Comment by josevalim** [comment]
I have reverted this PR for v1.6 but we can continue re-evaluating it.

@antipax clean up means it leaves no dangling processes behind.

**Pull request merged**